spec.bs

```diff
@@ -6,57 +6,882 @@ Status: CG-DRAFT
+  * {{SharedStorageRunOperation}}
+  * {{SharedStorageSelectURLOperation}}
+
+  A {{SharedStorageRunOperation}} is an {{SharedStorageOperation}} that has no JavaScript return value, and hence is especially suited for calling into the [=private aggregation=] service.
+
+  A {{SharedStorageSelectURLOperation}} is an {{SharedStorageOperation}} that takes in a [=list=] of {{SharedStorageUrlWithMetadata}}s (i.e. [=dictionaries=] containing [=strings=] representing {{URL}}s each wrapped with metadata) and returns a {{long}} integer index specifying which of these {{URL}}s should be selected.
+
+  <xmp class='idl'>
+    [Exposed=SharedStorageWorklet]
+    interface SharedStorageRunOperation : SharedStorageOperation {
+      Promise<undefined> run(object data);
+    };
+
+    [Exposed=SharedStorageWorklet]
+    interface SharedStorageSelectURLOperation : SharedStorageOperation {
+      Promise<long> run(object data,
+                        FrozenArray<SharedStorageUrlWithMetadata> urls);
+    };
+
+    dictionary SharedStorageUrlWithMetadata {
+      USVString url;
+      object reportingMetadata;
+    };
+  </xmp>
+
+  ## Registering a Worklet ## {#register-worklet}
+  <div algorithm>
+    For the <dfn method for="SharedStorageWorkletGlobalScope">register(|name|, |operationCtor|)</dfn> method, run the following steps [=in parallel=]:
+
+    1. If |name| is missing or empty, throw a {{TypeError}}.
+    1. Let |operationMap| be this {{SharedStorageWorkletGlobalScope}}'s [=SharedStorageWorkletGlobalScope/operation map=].
+    1. If |operationMap| [=map/contains=] an [=map/entry=] with [=map/key=] |name|, throw a {{TypeError}}.
+    1. If |operationCtor| is missing, throw a {{TypeError}}.
+    1. Let |isConstructor| be the result of running [=IsConstructor()=] with |operationCtor|.
+    1. If |isConstructor| is false, throw a {{TypeError}}.
+    1. Let |prototype| be the result of running |operationCtor|'s [=object/[[GetPrototypeOf]]()=] method.
+    1. If |prototype| is not an [=object=], throw a {{TypeError}}.
+    1. Let |run| be the result of running [=GetMethod()=] with |prototype| and "`run`".
+    1. If |run| is undefined, throw a {{TypeError}}.
+    1. [=map/Set=] the value of |operationMap|[|name|] to |run|.
+  </div>
+
+The [=Shared Storage Database=] {#database}
+===========================================
+A [=user agent=] holds a <dfn>shared storage database</dfn>, which consists of a [=shared storage shed=] and provides methods to [=shared storage database/store an entry in the database|store=], [=shared storage database/retrieve an entry from the database|retrieve=], [=shared storage database/delete an entry from the database|delete=], [=shared storage database/clear all entries in the database|clear=], and [=shared storage database/purge expired entries from the database|purge expired=] data, and additional methods as below. The data in the [=shared storage database|database=] take the form of [=shared storage database/entry|entries=].
```

Comment:
Here you say the user agent holds the database and the database holds the shed, but below you say the user agent holds the shed directly.  Which one is correct?

I think probably we want the user agent to hold the shed directly and then put the shared storage database inside one of the storage spec layers.  I think the storage bucket layer makes the most sense.  Its the layer where other storage APIs are defined, etc.  This would mean defining a new storage endpoint identifier:

https://storage.spec.whatwg.org/#registered-storage-endpoints

Which it looks like you do below with the string "shared".

Reply:
Changed to be somewhat similar to  https://html.spec.whatwg.org/multipage/webstorage.html#the-localstorage-attribute